src/java.base/share/classes/java/io/FileDescriptor.java

```diff
@@ -397,9 +422,62 @@ private synchronized void beforeCheckpoint() throws CheckpointOpenFileException
 
     private native String getType();
 
-    private synchronized void afterRestore() {
+    private native int getFlags();
+
+    private native long getOffset();
+
+    private synchronized void afterRestore() throws RestoreException {
+        if (!valid() && resource.originalFd >= 0) {
+            OpenFDPolicies.AfterRestorePolicy policy =
+                    OpenFDPolicies.RESTORE.get(resource.originalFd, resource.originalType, resource.originalPath);
+            if (policy.type == OpenFDPolicies.AfterRestore.KEEP_CLOSED) {
+                LoggerContainer.debug("FD %d (%s) is not reopened per policy%n",
+                            resource.originalFd, resource.originalPath);
+                resource.originalPath = null;
+                resource.originalType = null;
+                return;
+            }
+            String path;
+            if (policy.type == OpenFDPolicies.AfterRestore.OPEN_OTHER) {
+                path = policy.param;
+            } else {
+                if (resource.originalPath == null) {
+                    throw new RestoreException("Cannot reopen file descriptor " +
+                            resource.originalFd + ": invalid path: " + resource.originalPath);
+                } else if (resource.originalType.equals("socket")) {
+                    throw new RestoreException("Cannot reopen file descriptor " +
+                            resource.originalFd + ": cannot restore socket");
+                }
+                path = resource.originalPath;
+            }
+            // We will attempt to open at the original offset even if the path changed;
+            // this is used probably as the file moved on the filesystem but the contents
+            // are the same.
+            if (!reopen(resource.originalFd, path, resource.originalFlags, resource.originalOffset)) {
```

Comment:
Another option is a log file in the append mode, which may grow larger -- in that case we'd like to have that in append mode with position at the end. Probably that is handled by setting proper flags, but at least this would contradict with the comment.

Reply:
What I was trying to say that it's not about reading the 'append' flag when the file was opened, because that was just saying "let's open and then seek to the end of the file". We shouldn't conclude that we want to start at the 'new end' (provided that the file was modified) on restore.
Having a policy that will reopen at end, regardless of original offset, is certainly a valid option.